baseplate/ratelimit.py

```diff
@@ -0,0 +1,134 @@
+            raise RateLimitExceededException('Rate limit exceeded.')
+
+
+class RateLimitCache(object):
+    """An interface for rate limit backends to implement.
+
+    :param str key: The name of the rate limit bucket to consume from.
+    :param int amount: The amount to consume from the rate limit bucket.
+    :param int allowance: The maximum allowance for the rate limit bucket.
+    :param int interval: The interval to reset the allowance.
+
+    """
+    def consume(self, key, amount, max, bucket_size):
+        raise NotImplementedError
+
+
+class RedisRateLimitCache(RateLimitCache):
+    """A Redis-backed cache for rate limiting.
+
+    :param redis_client: An instance of
+        :py:class:`baseplate.context.redis.MonitoredRedisConnection`.
+
+    """
+
+    def __init__(self, redis_client):
+        self.redis_client = redis_client
+
+    def consume(self, key, amount, allowance, interval):
+        """Consume the given `amount` from the allowance for the given `key`.
+
+        This will return true if the `key` remains below the `allowance`
+        after consuming the given `amount`.
+
+        :param str key: The name of the rate limit bucket to consume from.
+        :param int amount: The amount to consume from the rate limit bucket.
+        :param int allowance: The maximum allowance for the rate limit bucket.
+        :param int interval: The interval to reset the allowance.
+
+        """
+        current_bucket = _get_current_bucket(interval)
+        key = key + current_bucket
+        ttl = interval * 2
+        with self.redis_client.pipeline('ratelimit') as pipe:
+            pipe.incr(key, amount)
+            pipe.expire(key, time=ttl)
+            responses = pipe.execute()
+        count = responses[0]
+        return count <= allowance
+
+
+class MemcacheRateLimitCache(RateLimitCache):
+    """A Memcache-backed cache for rate limiting.
+
+    :param memcache_client: An instance of
+        :py:class:`baseplate.context.memcache.MonitoredMemcacheConnection`.
+
+    """
+
+    def __init__(self, memcache_client):
+        self.memcache_client = memcache_client
+
+    def consume(self, key, amount, allowance, interval):
+        """Consume the given `amount` from the allowance for the given `key`.
+
+        This will return true if the `key` remains below the `allowance`
+        after consuming the given `amount`.
+
+        :param str key: The name of the rate limit bucket to consume from.
+        :param int amount: The amount to consume from the rate limit bucket.
+        :param int allowance: The maximum allowance for the rate limit bucket.
+        :param int interval: The interval to reset the allowance.
+
+        """
+        current_bucket = _get_current_bucket(interval)
+        key = key + current_bucket
+        ttl = interval * 2
+        self.memcache_client.add(key, 0, expire=ttl)
+        count = self.memcache_client.incr(key, 1)
```

Comment:
This will fail if there's a GC pause or something between these two memcache calls that lasts the duration of `ttl`

Reply:
It looks like keys must be set prior to using `incr`, and memcached doesn't have any sort of transactions. I'm not sure that this can be avoided, but maybe someone with more memcache experience knows of something.

Reply:
Looks like r2 actually handles this case in its ratelimit implementation. Maybe we should do something similar here.

Reply:
https://github.com/reddit-archive/reddit/commit/df3e8532149ac6d948a18c588a0e0aa52845e679

Reply:
yeah, it's best to catch it so the user doesn't get a crash. i think we can treat the ratelimits as best-effort though so being slightly off because of this sorta thing isn't the end of the world.

Reply:
Actually, the `pymemcache` docs say this case would return a `None` instead or raising an exception.
https://github.com/pinterest/pymemcache/blob/65560b738a01bfd420cdbeaae3b7bd78503c2855/pymemcache/client/base.py#L571

I think we should just treat it as zero in this case:

```
self.memcache.add(key, 0, expire=ttl)
count = self.memcache.incr(key, 1) or 0
return count <= allowance
```

Reply:
Actually, I guess `or amount` is more appropriate.